worker/src/direct_invocation/direct_invocation_ocall.rs

```diff
@@ -0,0 +1,153 @@
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+use crate::direct_invocation::watch_list_service::WatchList;
+use crate::direct_invocation::watching_client::WsWatchingClient;
+use crate::ocall_bridge::bridge_api::{
+    DirectInvocationBridge, OCallBridgeError, OCallBridgeResult,
+};
+use codec::{Decode, Encode};
+use sp_core::H256 as Hash;
+use std::sync::Arc;
+use substratee_worker_primitives::{DirectRequestStatus, RpcReturnValue, TrustedOperationStatus};
+use ws::CloseCode;
+
+/// implementation of the direct invocation o-call
+pub struct DirectInvocationOCall<W> {
+    watch_list: Arc<W>,
+}
+
+impl<W> DirectInvocationOCall<W> {
+    pub fn new(watch_list: Arc<W>) -> Self {
+        DirectInvocationOCall { watch_list }
+    }
+
+    fn continue_watching(status: &TrustedOperationStatus) -> bool {
+        !matches!(
+            status,
+            TrustedOperationStatus::Invalid
+                | TrustedOperationStatus::InSidechainBlock(_)
+                | TrustedOperationStatus::Finalized
+                | TrustedOperationStatus::Usurped
+        )
+    }
+}
+
+impl<W> DirectInvocationBridge for DirectInvocationOCall<W>
+where
+    W: WatchList<WsWatchingClient>,
+{
+    fn update_status_event(
+        &self,
+        hash_vec: Vec<u8>,
+        status_update_vec: Vec<u8>,
+    ) -> OCallBridgeResult<()> {
+        let status_update = TrustedOperationStatus::decode(&mut status_update_vec.as_slice())
+            .map_err(|e| OCallBridgeError::DirectInvocationError(format!("{:?}", e)))?;
+
+        let hash = Hash::decode(&mut hash_vec.as_slice())
+            .map_err(|e| OCallBridgeError::DirectInvocationError(format!("{:?}", e)))?;
+
+        let continue_watching = match self.watch_list.get_watching_client(&hash) {
+            Some(watched_client) => {
+                let old_result: Vec<u8> = watched_client.response.result.clone();
+                let mut result = RpcReturnValue::decode(&mut old_result.as_slice()).unwrap();
+
+                let do_watch = DirectInvocationOCall::<W>::continue_watching(&status_update);
+
+                // update response
+                result.do_watch = do_watch;
+                result.status = DirectRequestStatus::TrustedOperationStatus(status_update);
+
+                let mut updated_client = watched_client;
+                updated_client.response.result = result.encode();
+
+                updated_client
+                    .client
+                    .send(serde_json::to_string(&updated_client.response).unwrap())
+                    .unwrap();
```

Comment:
This unwrap I deem especially problematic. This panics if the ws-connection was closed by the client meanwhile, which can happen, as the client needs to wait for multiple seconds (max=currently layer1 blocktime=6s).

Reply:
Indeed, I changed that section of the code a bit - now we're returning Result types instead of panicking